PARQUET-485: Decouple page deserialization from column reader to facilitate unit testing

wesm opened the pull request:
Several things in this patch

* Adds PageReader abstraction, and a SerializedPageReader implementation
  according to the Parquet file format
* Adds a MockPageReader and a couple unit tests demonstrating end-to-end test
  without creating a Parquet file
* Adds a DataPageBuilder test fixture tool, may become part of the main write
  path later
* Adds PlainEncoder implementation for a few primitive types
* Fixes a few ColumnReader bugs exposed by the unit tests

wesm commented:
@majetideepak I'm going to add a couple test examples using the `MockPageReader` in this patch. If you could base your changes for PARQUET-169 on this it will also enable you to write unit tests with data that you declare with the test case. 

Testing will be slightly tricky as we do not yet have schema support for repeated fields, so you may wish to add a temporary member (the `max_repetition_level`) that you can set in the test case. I'm going to implement schema resolution this week so that the data type as well as the maximum repetition and definition levels are passed in to the column reader ctors. You can also wait for me to complete this work as it will reduce code / unit test churn.

yixinglu commented on src/parquet/column/reader-test.cc:
```diff
@@ -0,0 +1,84 @@
+    return;
+  }
+
+  out.resize(end - start);
```
`out.reserve(end - start);` ?

wesm replied:
Reserve doesn't change the array size

wesm replied:
Ooh there's actually a bug on line 65, will fix

julienledem commented on src/parquet/column/page.h:
```diff
@@ -0,0 +1,125 @@
+// semantics (the PageReader may or may not own the memory referenced by a
+// page)
+class Page {
+  // TODO(wesm): crc verification from parquet::PageHeader
```
FYI: parquet-mr never implemented crc verification. We had it in the spec but it is still a todo there too.

julienledem replied:
or generation

wesm replied:
Thanks, good to know!

julienledem commented:
Looks good to me.
Anybody else with comments?

wesm commented:
I'm going to do a little more on test cases. Will advise when done tomorrow

wesm commented:
OK, I've got a "hello world" test working on a REQUIRED field; I will add a test for an OPTIONAL field (with definition levels), then let's merge this. Hard to do a lot more without richer schema information. Code coverage will help steer further ColumnReader testing efforts

wesm commented:
OK, I added another unit test that verifies a flat optional field. It would be great if we can get this merged today as it simplifies unit testing `ColumnReader`. I fixed multiple correctness bugs (that would cause segfaults) as part of this patch.

We can continue to develop tooling to simplify setting up test fixtures and reducing the amount of boilerplate to a minimum.

wesm commented:
Just rebased this on master after #33 merged.

majetideepak commented:
I added test cases in #30 for the rle/bit-packed values encoding and decoding via the LevelEncoder and LevelDecoder API.

julienledem commented:
This looks good to me.
I'll merge it later today unless someone else has comments.

wesm commented:
I'd like @nongli to have a quick glance -- let me know.

nongli commented on src/parquet/encodings/encodings.h:
```diff
@@ -67,6 +67,38 @@ class Decoder {
+  // Subclasses should override the ones they support
+  //
+  // @returns: the number of bytes written to dst
+  virtual size_t Encode(const T* src, int num_values, uint8_t* dst) {
```
Are you sure you want dst to be a raw array instead of something like output stream? We can revisit later when the encoding side is built out more but this might be hard to deal with. Haven't put too much thought but maybe something blocky:
i.e.
byte* buffer = output->Reserve(big block for a batch).

wesm replied:
No, I'm not sure, I was also thinking about this because these classes have no ability to allocate memory. Let's create a JIRA to do that before we get in too deep with this assumption; I don't think this JIRA is the right place for it though...

wesm replied:
I created https://issues.apache.org/jira/browse/PARQUET-501 to track.

nongli commented on src/parquet/column/page.h:
```diff
@@ -0,0 +1,127 @@
+ public:
+  virtual ~PageReader() {}
+
+  // @returns: false on EOS, true if a new page has been returned
```
Should this just return shared_ptr<Page> and null on eos?

wesm replied:
Well, right now you can do

`if (!reader_->NextPage(&currentpage_))` which is nice. Are you suggesting returning `std::shared_ptr<PageReader>(nullptr)`?

wesm replied:
I thought some more about this and I'm fine with

```c++
current_page_ = reader->NextPage();
if (!current_page) {
  // eos
  ...
}
```

I wasn't sure whether returning a shared_ptr to null was standard practice but it seems fine to me.

wesm replied:
Let me know if that works for you

wesm replied:
Yeah, this is better, just pushed the change.

nongli commented on src/parquet/column/serialized-page.h:
```diff
@@ -0,0 +1,61 @@
+namespace parquet_cpp {
+
+// This subclass delimits pages appearing in a serialized stream, each
+// proceeded by a serialized Thrift parquet::PageHeader indicating the type of
```
preceded

wesm replied:
thanks will fix

nongli commented on src/parquet/column/page.h:
```diff
@@ -0,0 +1,127 @@
+
+namespace parquet_cpp {
+
+// Note: Copying the specific page header Thrift metadata to the Page
```
I'm not sure I understand this comment.

wesm replied:
Well, the issue here is that the deserialized page header has a limited lifetime in the current implementation. If you wanted to pipeline page processing, you need some way to preserve the page metadata for all of the "live" (unprocessed) pages that have been buffered in memory.

wesm replied:
I.e., I'm referring to the `current_page_header_` member variable (which gets overwritten each time a page header is deserialized)

wesm replied:
Another option is to have each of these classes have a pointer to metadata, and if you want to have some pipelining abstraction then you can create a container that owns the deserialized metadata to eliminate unnecessary copies.

wesm commented:
Just rebased and restored a missing function that got lost in the RLE sync up. This should be good to go when the build is green.

wesm commented:
Travis CI is super unhappy today, seems like a trend. I sent them a support email about it

wesm commented:
:100: thank you! rebasing #34 now
